test/functional/mempool_reorg.py

```diff
@@ -108,6 +108,37 @@ def run_test(self):
         assert_equal(set(self.nodes[0].getrawmempool()), set())
         self.sync_all()
 
+        self.log.info("Simulate a natural reorg where mempool transaction stays valid but on a different block")
+        self.connect_nodes(0, 1)
+        # Mine a block to create mature coinbase
```

Comment:
This test passes also if the fix is missing. Is this expected? If yes, then what is the point of the test? I think tests should fail if the bug resurfaces.

On top of this PR:
* `git show b4adc5ad67 |git apply -R`
* recompile
* run `test/functional/mempool_reorg.py` - it passes

Reply:
I don't think it is possible to test this without modifying the source code (e.g. addding an assert) or inspecting the process internal memory.

Reply:
So, what is the purpose of this test? It should be removed?

Reply:
I think the goal is to do a reorg without using `invalidateblock`, as above

Reply:
The test is supposed to fail without the bugfix if the asserts are there

Reply:
I recall adding the assert and saw this and one other test failing already

Reply:
What assert? Something extra, that is not in the PR?

Reply:
Yes, assert.

As mentioned previously it is not possible to test this without modifying the source code or looking at internal memory.

Adding an assert is one way to modify the source code. Another way would be to return the struct over rpc, or log it to the debug log, or ...

Reply:
> What assert?

This one: https://github.com/bitcoin/bitcoin/pull/23683/files#diff-c065d4cd2398ad0dbcef393c5dfc53f465bf44723348892395fffd2fb3bac522R642

Reply:
In a previous version of this PR I think I had added the assert in a separate commit from the bugfix.

Reply:
I see (which assert). So I took latest `master` (7629efcc2c3a8a8a7c17b1300cd466ec6c8c1f3f) and applied this patch:

```diff
--- i/src/txmempool.cpp
+++ w/src/txmempool.cpp
@@ -653,6 +653,7 @@ void CTxMemPool::removeForReorg(CChain& chain, std::function<bool(txiter)> check
         if (!TestLockPointValidity(chain, lp)) {
             mapTx.modify(it, update_lock_points(lp));
         }
+        assert(TestLockPointValidity(chain, lp));
     }
 }
```

`test/functional/mempool_reorg.py` then fails because the added assert is hit. This is unmodified `mempool_reorg.py`, without the extension from this PR. If the bug can be reproduced by adding an assert + unmodified `mempool_reorg.py`, then why bother to change (extend) the test?

> I think the goal is to do a reorg in this test without using `invalidateblock`

Why bother if the test fails even without that?

Reply:
reorgs in the real network happen without `invalidateblock`, so I think we should attempt to not use the call

Reply:
The new test is not strictly necessary in order to hit this, but it's not pointless. We didn't have any natural reorgs. We used `invalidateblock` as a shortcut to do it manually. I can remove it if you insist, but I don't really see why it's problematic?

Reply:
Alright, I agree that for the test the natural reorg is better than `invalidateblock`. But the change in this PR just adds a natural reorg, without removing the `invalidateblock` one. Shouldn't it replace `invalidateblock` with a natural reorg?

I checked the code coverage with and without the extension of the test from this PR (28b60ce3122b29ffad2c60790d3bfe9a2ede2471) and it is the same. In other words, the newly added snippet does not extend the coverage. It exercises code that has already been exercised by preceding parts of `mempool_reorg.py`.

<details>
<summary>commands to compare coverage with/without test extension</summary>

```sh
# compile
F="-fprofile-instr-generate -fcoverage-mapping" CXXFLAGS=${F} LDFLAGS=${F} ./configure and make as usual

# run test and gather coverage data
LLVM_PROFILE_FILE="/tmp/coverage1/%m-%p.profraw" test/functional/mempool_reorg.py
llvm-profdata13 merge /tmp/coverage1/*.profraw -o /tmp/coverage1/all.profdata

# remove topmost commit (test extension)
git show |git apply -R

# run test and gather coverage data
LLVM_PROFILE_FILE="/tmp/coverage2/%m-%p.profraw" test/functional/mempool_reorg.py
llvm-profdata13 merge /tmp/coverage2/*.profraw -o /tmp/coverage2/all.profdata

# compare the coverage (both are identical)
md5 -r /tmp/coverage*/all.profdata
14196d89227bc6028834aeff32b79ff5 /tmp/coverage1/all.profdata
14196d89227bc6028834aeff32b79ff5 /tmp/coverage2/all.profdata
```
</details>

I don't insist on removing the test snippet. I am just trying to understand the point of it. So far, with my understanding, I am ~0 on it because I fail to see what value it adds. If the `invalidate` block reorg is removed and replaced by a natural reorg without reducing code coverage, then that would be a definite +1.